ChatBot/SteamWorks/LobbyWatcher.cs

```diff
@@ -0,0 +1,448 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.NetworkInformation;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using SteamKit2;
+
+#nullable enable
+
+namespace ChatBot.SteamWorks
+{
+    public enum WatcherState
+    {
+        Invalid = 0,
+
+        Connecting,
+        Connected,
+        LoggingIn,
+        Received2FARequest,
+        Submitting2FAResponse,
+        LoggedIn,
+        Searching,
+        Idle,
+        Disconnecting,
+
+        Completed,
+        Disconnected,
+        LoginFailed,
+        Errored,
+        Disposed,
+    }
+
+    // Short-lived state if we get disconnected due to 2FA.
+    public class Steam2FAData
+    {
+        public string? AppCode { get; set; } = null;
+        public string? EmailCode { get; set; } = null;
+    }
+
+    public class LobbyWatcher : IDisposable
+    {
+        public Task? PollingTask { get; private set; } = null;
+
+        public WatcherState Status { get; private set; } = WatcherState.Invalid;
+
+
+        private readonly Steam2FAData _twoFAdata;
+        private readonly List<IDisposable> _cbHandles;
+        private readonly CallbackManager _cbManager;
+        private readonly Action<IEnumerable<ThunderDomeLobby>> _matchAction;
+        private readonly Func<ThunderDomeLobby, bool> _matchPredicate;
+        private readonly SteamClient _steamClient;
+
+        private uint? _cellId;
+
+        private SteamMatchmaking MatcherImpl => _steamClient.GetHandler<SteamMatchmaking>()
+            ?? throw new InvalidCastException("Failed to retrieve SteamMatchmaking interface");
+
+        private SteamUser UserImpl => _steamClient.GetHandler<SteamUser>()
+            ?? throw new InvalidCastException("Failed to retrieve SteamUser interface");
+
+
+        public LobbyWatcher(Action<IEnumerable<ThunderDomeLobby>>? matchAction = null, Func<ThunderDomeLobby, bool>? matchPredicate = null)
```

Comment:
I ended up getting rid of matchAction for now because logging to chat/a front end UI is the goal rather than logging to console (I think I see why you did it this way, its a new style to me and pretty elegant)

Reply:
I did add it back cause it allows for separation from UI layer and the LobbyWatcher